plugins/flashrom/fu-flashrom-lspcon-i2c-spi-device.c

```diff
@@ -0,0 +1,315 @@
+	 * run-time data */
+	struct romentry *entries;
+	/* the number of successfully parsed entries */
+	size_t num_entries;
+};
+
+static struct romentry entries[3] = {
+	{
+		.start = 0x00002,
+		.end = 0x00003,
+		.included = true,
+		.name = "FLAG"
+	},
+	{
+		.start = 0x10000,
+		.end = 0x1ffff,
+		.included = true,
+		.name = "PARTITION1"
+	},
+	{
+		.start = 0x20000,
+		.end = 0x2ffff,
+		.included = true,
+		.name = "PARTITION2"
+	},
+};
+
+static struct flashrom_layout layout = {entries, 3};
+
+#define VERSION_ADDR_OFFSET	0x5000
+
+
+static void
+fu_flashrom_lspcon_i2c_spi_device_init (FuFlashromLspconI2cSpiDevice *self)
+{
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_version_format (FU_DEVICE (self), FWUPD_VERSION_FORMAT_PAIR);
+}
+
+static gboolean
+fu_flashrom_lspcon_i2c_spi_device_prepare (FuDevice *device,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
+	FuFlashromLspconI2cSpiDevice *self = FU_FLASHROM_LSPCON_I2C_SPI_DEVICE (device);
+
+	/* flashrom_programmer_init() mutates the programmer_args string. */
+	fu_flashrom_device_set_programmer_args (parent,
+						g_strdup_printf	("bus=%d", self->bus_number));
+
+	/* FuFlashromDevice->prepare */
+	if (!FU_DEVICE_CLASS
+	    (fu_flashrom_lspcon_i2c_spi_device_parent_class)->prepare (device,
+								       FWUPD_INSTALL_FLAG_NONE,
+								       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gint
+fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (const gchar *path,
+							    GError **error)
+{
+	gint bus_number = -1;
+	GRegex *regex = NULL;
+	GMatchInfo *info;
+
+	regex = g_regex_new (I2C_PATH_REGEX, 0, 0, error);
+	if (regex == NULL)
+		return FALSE;
+
+	if (g_regex_match_full (regex, path, -1, 0, 0, &info, error)) {
+		bus_number = g_ascii_strtoll ( g_match_info_fetch (info, 1),
+					       NULL, 10);
+	}
+
+	g_match_info_free (info);
+	g_regex_unref (regex);
+
+	return bus_number;
+}
+
+static gboolean
+fu_flashrom_lspcon_i2c_spi_device_setup (FuDevice *device, GError **error)
+{
+	FuFlashromDevice *parent = FU_FLASHROM_DEVICE (device);
+	FuFlashromLspconI2cSpiDevice *self = FU_FLASHROM_LSPCON_I2C_SPI_DEVICE (device);
+	const gchar *path = NULL;
+	const gchar *hw_id = NULL;
+	g_autofree gchar *vid = NULL;
+	g_autofree gchar *pid = NULL;
+	g_autofree gchar *guid = NULL;
+	g_autofree gchar *contents = NULL;
+	struct flashrom_flashctx *flashctx = NULL;
+	gsize flash_size;
+	uint32_t addr;
+
+	if (g_strcmp0 (fu_flashrom_device_get_programmer_name (parent),
+		       "lspcon_i2c_spi") != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid programmer");
+		return FALSE;
+	}
+
+	path = fu_udev_device_get_sysfs_path (FU_UDEV_DEVICE (device));
+	self->bus_number = fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path (path, error);
+	if (self->bus_number == -1) {
+		g_set_error (error,
```

Comment:
isn't error going to be set by `fu_flashrom_lspcon_i2c_spi_device_get_bus_number_from_path`

Reply:
not in every case, but I can move this above

Reply:
I think it would be better to follow the style of having -1 mean error and set error properly in that function.

Reply:
I'd much rather have the function return a gboolean TRUE/FALSE and then pass &self->bus_number as an out parameter...

Reply:
I moved the logic into ->probe() as it was short enough..